### PR TITLE
Add round trip tests for reading/writing parquet metadata

### PR DESCRIPTION
**Pull request by alamb**
Draft as this PR builds on https://github.com/apache/arrow-rs/pull/6466

# Which issue does this PR close?

Part of https://github.com/apache/arrow-rs/issues/6002
# Rationale for this change
 
@adriangb , @etseidl  and myself were unclear what happens when parquet metadata that has been serialized directly with the ParquetMetaDataWriter: https://github.com/apache/arrow-rs/pull/6081/files#r1773996251

I figured the best way to find out was to write some tests for it 🤓 

# What changes are included in this PR?

# Are there any user-facing changes?
No, just tests

### CODE REVIEW

**Comment by adriangb** [comment]
I think this PR or a subsequent one can also then delete this horrid tests: https://github.com/apache/arrow-rs/blob/2cc0c167d9725c1b1965980965d637928d887380/parquet/src/file/metadata/writer.rs#L416-L676

**Comment by alamb** [comment]
> I think this PR or a subsequent one can also then delete these tests that I wrote before we had the dedicated reader:

Good idea, I did it in 904596859b

File: parquet/src/file/metadata/writer.rs [hunk 1]
```diff
@@ -377,300 +377,3 @@ impl<'a, W: Write> ParquetMetaDataWriter<'a, W> {
         }
     }
 }
-
-#[cfg(test)]
```
Comment (alamb) [inline parquet/src/file/metadata/writer.rs:381]: Removed per @adriangb 's comment: https://github.com/apache/arrow-rs/pull/6463#issuecomment-2377345221

File: parquet/src/arrow/mod.rs [hunk 2]
```diff
@@ -236,3 +236,146 @@ pub fn parquet_column<'a>(
...
+            .unwrap();
+
+        assert_eq!(original_metadata, roundtrip_metadata);
+    }
+
+    #[test]
+    fn test_metadata_read_write_roundtrip_page_index() {
+        let parquet_bytes = create_parquet_file();
+
+        // read the metadata from the file including the page index  structures
```
Comment (mbrobbel) [inline parquet/src/arrow/mod.rs:277]:
```suggestion
        // read the metadata from the file including the page index structures
```

File: parquet/src/arrow/mod.rs [hunk 3]
```diff
@@ -236,3 +236,146 @@ pub fn parquet_column<'a>(
...
+
+        let mut writer = ArrowWriter::try_new(&mut buf, batch.schema(), Some(props)).unwrap();
+        writer.write(&batch).unwrap();
+        writer.finish().unwrap();
+        drop(writer);
+
+        Bytes::from(buf)
+    }
+
+    /// Serializes `ParquetMetaData` into a memory buffer, using `ParquetMetadataWriter
```
Comment (mbrobbel) [inline parquet/src/arrow/mod.rs:349]:
```suggestion
    /// Serializes `ParquetMetaData` into a memory buffer, using `ParquetMetadataWriter`
```

File: parquet/src/arrow/mod.rs [hunk 4]
```diff
@@ -236,3 +236,146 @@ pub fn parquet_column<'a>(
...
+    fn metadata_to_bytes(metadata: &ParquetMetaData) -> Bytes {
+        let mut buf = vec![];
+        ParquetMetaDataWriter::new(&mut buf, metadata)
+            .finish()
+            .unwrap();
+        Bytes::from(buf)
+    }
+
+    /// Sets the page index offset locations in the metadata None
```
Comment (mbrobbel) [inline parquet/src/arrow/mod.rs:358]:
```suggestion
    /// Sets the page index offset locations in the metadata to `None`
```

**Review by mbrobbel — APPROVED** [review]

**Review by alamb — COMMENTED** [review]
Thank you @mbrobbel for the review

**Comment by alamb** [comment]
Thanks again for the help and review @adriangb  and @mbrobbel

**Pull request merged**